vortex-serde/src/layouts/write/metadata_accumulators.rs

```diff
@@ -0,0 +1,263 @@
+        DType::Bool(..) => Box::new(ExtremaAccumulator::<bool>::new(hint, into_bool_array)),
+        DType::Primitive(ptype, ..) => {
+            match_each_native_ptype!(ptype, |$P| {
+                Box::new(ExtremaAccumulator::<$P>::new(hint, into_primitive_array::<$P>))
+            })
+        }
+        DType::Utf8(..) => Box::new(ExtremaAccumulator::<BufferString>::new(
+            hint,
+            into_utf8_array,
+        )),
+        DType::Binary(..) => Box::new(ExtremaAccumulator::<Buffer>::new(hint, into_binary_array)),
+        DType::Struct(..) => Box::new(BasicAccumulator::new(hint)),
+        DType::List(..) => Box::new(BasicAccumulator::new(hint)),
+        DType::Extension(..) => Box::new(BasicAccumulator::new(hint)),
+    }
+}
+
+pub trait MetadataAccumulator {
+    fn push_chunk(&mut self, array: &Array) -> VortexResult<()>;
+
+    fn push_batch_byte_offsets(&mut self, batch_byte_offsets: Vec<u64>);
+
+    fn into_layouts_and_metadata(self: Box<Self>) -> VortexResult<(VecDeque<Layout>, StructArray)>;
+
+    fn into_layouts_and_metadata_parts(
+        self,
+    ) -> VortexResult<(VecDeque<Layout>, Vec<String>, Vec<Array>)>;
+}
+
+struct ExtremaAccumulator<T> {
+    minima: Vec<Option<T>>,
+    maxima: Vec<Option<T>>,
+    to_array: fn(Vec<Option<T>>) -> Array,
+    basic_metadata: Box<BasicAccumulator>,
+}
+
+impl<T> ExtremaAccumulator<T> {
+    fn new(size_hint: usize, to_array: fn(Vec<Option<T>>) -> Array) -> Self {
+        Self {
+            minima: Vec::with_capacity(size_hint),
+            maxima: Vec::with_capacity(size_hint),
+            to_array,
+            basic_metadata: Box::new(BasicAccumulator::new(size_hint)),
+        }
+    }
+}
+
+impl<T> MetadataAccumulator for ExtremaAccumulator<T>
+where
+    T: TryFrom<Scalar, Error = VortexError>,
+{
+    fn push_chunk(&mut self, array: &Array) -> VortexResult<()> {
+        self.minima.push(
+            array
+                .statistics()
+                .compute(Stat::Min)
+                .map(T::try_from)
+                .transpose()?,
+        );
+        self.maxima.push(
+            array
+                .statistics()
+                .compute(Stat::Max)
+                .map(T::try_from)
+                .transpose()?,
+        );
+
+        self.basic_metadata.push_chunk(array)
+    }
+
+    fn push_batch_byte_offsets(&mut self, batch_byte_offsets: Vec<u64>) {
+        self.basic_metadata
+            .push_batch_byte_offsets(batch_byte_offsets);
+    }
+
+    fn into_layouts_and_metadata(self: Box<Self>) -> VortexResult<(VecDeque<Layout>, StructArray)> {
+        let (chunks, names, fields) = self.into_layouts_and_metadata_parts()?;
+        let n_chunks = chunks.len();
+        let names = fieldnames_from_strings(names);
+        Ok((
+            chunks,
+            StructArray::try_new(names, fields, n_chunks, Validity::NonNullable)?,
+        ))
+    }
+
+    fn into_layouts_and_metadata_parts(
```

Comment:
This is never called, right? Can make `into_layouts_and_metadata_parts` a fn on `BasicAccumulator` instead of the trait

Reply:
Done.